contracts/src/v0.8/tests/UpkeepCounterStats.sol

```diff
@@ -47,65 +167,122 @@ contract UpkeepCounterStats {
 
   function performUpkeep(bytes calldata performData) external {
     uint256 startGas = gasleft();
+    (uint256 upkeepId, ) = abi.decode(
+      performData,
+      (uint256, bytes)
+    );
+    uint256 performGasToBurn = upkeepIdsToPerformGasToBurn[upkeepId];
+    uint256 initialBlock = upkeepIdsToInitialBlock[upkeepId];
+    uint256 blockNum = block.number;
+    uint256 interval = upkeepIdsToIntervals[upkeepId];
     if (initialBlock == 0) {
-      initialBlock = block.number;
+      upkeepIdsToInitialBlock[upkeepId] = blockNum;
+      initialBlock = blockNum;
     } else {
       // Calculate and append delay
-      uint256 delay = block.number - previousPerformBlock - interval;
-      delays.push(delay);
+      uint256 delay = blockNum - upkeepIdsToPreviousPerformBlock[upkeepId] - interval;
+      upkeepIdsToDelay[upkeepId].push(delay);
     }
 
-    lastBlock = block.number;
-    counter = counter + 1;
-    emit PerformingUpkeep(initialBlock, lastBlock, previousPerformBlock, counter, performData);
-    previousPerformBlock = lastBlock;
+    uint256 counter = upkeepIdsToCounter[upkeepId] + 1;
+    upkeepIdsToCounter[upkeepId] = counter;
+    emit PerformingUpkeep(initialBlock, blockNum, upkeepIdsToPreviousPerformBlock[upkeepId], counter, performData);
+    upkeepIdsToPreviousPerformBlock[upkeepId] = blockNum;
+
+    // every upkeep adds funds for themselves
+    if (blockNum - upkeepIdsToLastTopUpBlock[upkeepId] > upkeepTopUpCheckInterval) {
+      UpkeepInfo memory info = registry.getUpkeep(upkeepId);
+      uint96 minBalance = registry.getMinBalanceForUpkeep(upkeepId);
+      if (info.balance < minBalanceThresholdMultiplier * minBalance) {
+        this.addFunds(upkeepId, addFundsMinBalanceMultiplier * minBalance);
+        upkeepIdsToLastTopUpBlock[upkeepId] = blockNum;
+        emit UpkeepTopUp(upkeepId, addFundsMinBalanceMultiplier * minBalance, blockNum);
+      }
+    }
```

Comment:
Let's simplify this by deploying a new registry with mock gas feed. then we no longer need to worry about link funds